Feature/#2 init data setup

Yuurim98 opened the pull request:
## ✨ 어떤 기능인가요?

<!-- 구현한 기능에 대해 간단히 설명해 주세요 -->
- DB 설정 (h2 사용) 및 더미 데이터 추가, 엔티티 정의

---

## ✅ 작업 내용 체크리스트

<!-- 작업한 항목을 체크박스로 작성해 주세요 -->
- [ ] 기능 구현
- [ ] 단위 테스트
- [ ] 코드 리팩토링
- [ ] 문서 작성
- [x] 초기 설정

---

## 📸 스크린샷 (선택)

<!-- UI 변경이 있다면 첨부해 주세요 -->

---

## 🔗 관련 이슈

<!-- 이 PR과 연결된 이슈가 있다면 연결해 주세요 -->
- close #2

yu-rim98 committed:
Remove: properties 삭제

yu-rim98 committed:
Env: yml 파일 생성 및 설정 #2

더미 데이터 sql 파일 추가

yu-rim98 committed:
Feat: 엔티티 생성 #2

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/VelogCloneApplication.java:
```diff
@@ -2,8 +2,9 @@
+import org.springframework.boot.autoconfigure.security.servlet.SecurityAutoConfiguration;
 
-@SpringBootApplication
+@SpringBootApplication(exclude = SecurityAutoConfiguration.class)
```
오, 이건 왜 추가하신 건지 궁금합니다. security를 사용하신 건가요? 아니면, security를 사용하지 않아도 사용할 이유가 있나요?

Yuurim98 replied:
security를 사용할 예정이라 의존성을 추가해두었는데 아직 보안 설정은 하지 않은 상태라 임시로 설정해두었습니다 (h2 콘솔로 접속해서 데이터가 삽입됐는지 확인하려는 용도도 있었음)

Sooamazing replied:
오 저 잘 모르는데, H2 콘솔로 접속해서 데이터 삽입됐는지 확인하려는 용도는 뭔가요? 
security를 사용하기 때문에 콘솔 접속이 어려울 수 있다는 건가요? 
그니까, 시큐리티를 사용하지 않았다면 H2 콘솔 접속 그냥 가능하다?

Yuurim98 replied:
지금 브라우저로 h2 콘솔 접속해서 DB 데이터를 확인하고 있는데 security 의존성을 추가하면 자동으로 기본 보안 ? 이 설정돼서 접속할 때마다 로그인을 해줘야 해서 비활성화 처리해두었어요.
질문주신게 이게 맞는지 모르겠네요

Sooamazing replied:
저 궁금한 게 생겼습니다. 저희 session 로그인으로 간단하게 구현하기로 한 거 맞나요? security를 이용한다는 게 그런 의미인가요?

Yuurim98 replied:
security를 활용하지 않고도 session 로그인은 구현할 수 있을 것 같습니다.
저는 security를 활용해서 구현할 생각으로 추가했었는데 처음에는 없이 구현하고 나중에 추가해서 변경해 보는 방식도 좋을 것 같아서 이렇게 진행할 수도 있을 것 같아요.

Sooamazing replied:
아하 넵. security 없이 구현할 때 salt, hashing 고민해 보시면 좋을 것 같아요.(제가 고민하는 중이라 말해 봄)

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/common/entity/BaseEntity.java:
```diff
@@ -0,0 +1,25 @@
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
+
+    protected LocalDateTime deletedAt;
```
(P3-답 궁금.) deletedAt까지 상속하신 이유가 궁금합니다. 모든 곳에서 soft Delete를 택하신 건가요? 왜 모든 곳에 공통으로 soft delete를 적용하셨는지 궁금합니다.
예를 들어, 팔로우나 좋아요 같은 것들은 취소한 날짜를 보관할 가치가 있는지, 있다면, 어떤 곳에 이용할 수 있을지 궁금합니다.

Yuurim98 replied:
아! hard delete로 처리할 것 같아서 제외해도 될 것 같습니다 (삭제 정책을 지금 생각했습니다 ㅋㅋ)

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/comment/entity/Comment.java:
```diff
@@ -0,0 +1,44 @@
+    @OneToMany(mappedBy = "parent")
+    private List<Comment> children = new ArrayList<>();
+
+    @Lob
+    @Column(nullable = false)
+    private String content;
```
(P3-답 궁금)
Lob를 사용하신 이유가 궁금합니다. 전 한 번도 사용해 보지 않았는데, 찾아보니, 이렇더군요. (참고: [@Lobil columnDefinition = "TEXT"](https://yeongjin.tistory.com/157) - 공식문서까진 못 봄.)
> TEXT는 MySQL 등 일부 데이터베이스에서 사용되는 데이터 타입으로, 긴 문자열을 저장하기 위한 타입입니다. TEXT 타입은 일반적으로 최대 65,535자까지 저장할 수 있습니다.

의도하신 게 최대 65,535자가 맞다면, 댓글인데, 65,535자까지 저장할 필요가 있을까요?

Yuurim98 replied:
벨로그가 댓글 제한이 없는 것 같아서 우선 TEXT 타입으로 지정했는데 의견주신 것처럼 댓글은 굳이 해당 타입으로 저장할 필요는 없는 것 같습니다 감사합니다!

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/comment/entity/Comment.java:
```diff
@@ -0,0 +1,44 @@
+    @JoinColumn(name = "post_id")
+    private Post post;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "comment_id")
+    private Comment parent;
+
+    @OneToMany(mappedBy = "parent")
+    private List<Comment> children = new ArrayList<>();
```
(P3-답 궁금.) 이건 댓글의 깊이 구현을 위한 건가요? 그럼, 댓글은 parent, children  양 쪽을 다 가지고 있는 게 맞나요? 맞다면, 양쪽을 갖고 있도록 설계하신 이유가 궁금합니다. 순환 참조 같은 건 딱히 안 날까요?(몰라서 물음.)

Yuurim98 replied:
댓글 자신이 특정 댓글의 대댓글일 수 있고 동시에 대댓글들을 가질 수도 있다고 생각해서 이렇게 설계했습니다. 
순환 참조 문제를 제가 겪은 적이 없었던 것 같아서 우선 이렇게 진행하다가 문제가 생기면 해결하는 것도 좋을 것 같아서 그 부분은 일단 고려하지 않았습니다.

Sooamazing replied:
오 순환 참조 나면 꼭 공유해 주세요. 궁금.

Sooamazing commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,19 @@
+    url: jdbc:h2:mem:testdb;MODE=MYSQL
+    username: sa
+    password:
+  h2:
+    console:
+      enabled: true
+      path: /h2-console
```
(P4-답 안 해도 됨. 근데 궁금.) h2 선택하신 이유가 궁금합니다. 전 h2 써 볼까 했는데, 잘 몰라서 MySQL 선택했거든요.

Yuurim98 replied:
이번 프로젝트는 배포 예정이 없기 때문에 로컬 환경에서만 실행되는 걸 전제로 비교적 데이터베이스 설정이 간단하고 초기화가 쉬운 H2 인메모리 데이터베이스를 선택했습니다. 따로 데이터베이스 서버를 띄우지 않아도 되는 장점도 있는 것 같습니다

Sooamazing commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,19 @@
+  datasource:
+    driver-class-name: org.h2.Driver
+    url: jdbc:h2:mem:testdb;MODE=MYSQL
+    username: sa
+    password:
```
(P3-답궁금) h2는 이렇게 username, password가 만천하에 공개되어도 상관 없나요? 이게.. 그.. 인메모리였나 그거여서 상관 없나요...? 해커가 막 이걸 보고 침입 시도할 일이 없나요?

Yuurim98 replied:
보통 환경 별로 yml 파일을 따로 관리하기도 하고 배포하면 h2는 사용되지 않으니까 상관 없지 않을까요.. 흠 한번도 생각 안 해봤던 건데 궁금하긴 하네요

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/follow/entity/Follow.java:
```diff
@@ -0,0 +1,27 @@
+    private User follower;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "followee_id")
```
헐. followee라고도 쓰는군요....! ㅋㅋㅋ 전 그냥 Following이라 해서 신기하네요.

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/user/entity/User.java:
```diff
@@ -0,0 +1,40 @@
+
+@Entity
+@Table(name = "USERS")
+public class User extends BaseEntity {
```
(P5-그냥 궁금) 오 전 user 이름이 궁금합니다. 전 항상 Member와 User 사이에서 고민하곤 해서요. 테이블 명을 USERS라고 준 것도, 예약어 때문이죠? 그래서 궁금합니다.

Yuurim98 replied:
예약어 때문에 별도로 지정해줬습니다
저도 Member랑 User 중에 고민을 하는데 Member 보다는 User가 더 명확? 한 느낌......

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/post/entity/Post.java:
```diff
@@ -0,0 +1,55 @@
+    @JoinColumn(name = "series_id")
+    private Series series;
+
+    @Column(nullable = false)
```
(P3-답 궁금)
nullable로 사용하신 이유가 궁금합니다.
저 사실 다 까먹어서 검색했는데(ㅋㅋ), NotNull, nullable로 보통 쓰더라구요. 근데, nullable은 어플리케이션 단에서(코드에서) Not null을 보장해 주지 않는 것 같아서 따로 서비스/도메인 코드에 구현하실 계획이신 건지 궁금합니다.
추가로, 제가 본 [블로그1](https://dd-developer.tistory.com/116), [블로그2](https://kafcamus.tistory.com/15)에서는 (검증은 안 했지만) `spring.jpa.properties.hibernate.check_nullability=true` 이걸 쓰면 nullable도 코드에서 not null을 보장해 준다고 합니다. 근데, 둘은 적용 시기는 다르대요. (거기까진 제대로 못 봄.)

Yuurim98 replied:
요청을 받을 때 Validation 처리할 생각이었는데 공유해주신 내용 보니까 NotNull 어노테이션을 달면 DDL 생성 시 제약 조건 추가, 엔티티 생성 시점에 검증도 되네요. 처음 알았습니다

궁금한게 있는데 잘은 모르지만 보통 DTO에서 Validation 처리만 하는 것 같은데 (제가 그랬음) 별도로 엔티티 객체 자체에서도 검증을 해야 한다고 보시나요 ? (쓰다보니 그래야 할 것 같네요 그래도 의견주시면 감사하겠습니다)

Sooamazing replied:
DTO에서 validation 처리라고 말씀하시는 거면, request 시의 dto를 말씀하시는 게 맞나요?

1. 맨 처음 요청 시 dto에서만 검증 
2. db 스키마 제약만 
3. 둘 다 사용

 이렇게 세 가지 경우가 있다고 할 때, 전 둘 다 사용할 것 같긴 합니다. 그런데, 저도 정확히 왜 그래야 하는지는 모르겠네요. (SQL을 시스템에서 따로 보내는 경우..... 실수하지 않으려고인가..? 그럴 일이 얼마나 있을지 모르겠지만....)
 질문 주신 의도도 아마 둘 다 하는 걸 의도하셨을 것 같긴 한데, 무튼 최대한 어플리케이션 코드에서 검증을 하고(라기엔 입력 받을 때 NotNull 어노테이션이면 되지만...), DB에 접근하도록 코드를 짤 것 같습니다.

 그리고, DB에 접근할 때는 가급적 DB에서 예외 터지지 않게 하는 게 코드에서 검증하는 게 어디서 본 것 같아요. (예외 터지면, 그걸 처리하고 또 접근하는 게 리소스가 들기 때문에 두 배로 리소스를 사용하기 때문이라고 본 것 같습니다. 락과도 관련이 있는지는 모르겠는데, 뭐 이런저런 이유가 더 있을 수도.)

별개로 꼭 필요한 경우가 아니면 성능 때문에 FK는 코드에서만 걸고, 실제 스키마에서는 안 건다는 글을 꽤 봤는데요. not null 설정도 동일할지 궁금하긴 합니다.
 [이 글](https://jojoldu.tistory.com/718)은 Not null 제약 추가 여부에 대한 글이긴 한데, 이런 걸 보면 sql 쿼리 날릴 때 Number, Boolean은 Not null을 스키마에 제약 조건을 거는 자체가 중요해 보이긴 합니다. (다른 건 여전히 조금 의문이지만.)

Yuurim98 replied:
> 그리고, DB에 접근할 때는 가급적 DB에서 예외 터지지 않게 하는 게 코드에서 검증하는 게 어디서 본 것 같아요. (예외 터지면, 그걸 처리하고 또 접근하는 게 리소스가 들기 때문에 두 배로 리소스를 사용하기 때문이라고 본 것 같습니다. 락과도 관련이 있는지는 모르겠는데, 뭐 이런저런 이유가 더 있을 수도.)

저는 잘 모르는 부분이긴 하지만 일리가 있는 말 같습니다. 이해가 조금 된 것 같아요 좋은 의견주셔서 감사합니다.

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/post/entity/Post.java:
```diff
@@ -0,0 +1,55 @@
+
+    private String content;
+
+    private String postIntro;
```
(P5 - 그냥 궁금)
오 이것도... 필수인가요? 소개는 안 적어도 되는 거라고 생각했는데, 필수로 하신 건가요? 
아니면, 사용자가 적지 않으면, 코드에서 잘라서 넣어줄 예정인가요? 그럼 나중에 수정할 때 불러오는 값이 달라지겠군요.

Yuurim98 replied:
필수라는게 어떤 걸 말씀하시는지 잘 모르겠습니다..
소개를 안 적으면 150자를 DB에 저장하려고 했는데 지금 생각해보니 그냥 null로 두거나 빈 문자열을 넣고 그대로 내려주면 어떨까 싶습니다. 프런트에서 글 내용 150자 뽑아서 보여주는 건 불가능 할까요? 아니면 내려주기 전에 필드에 담아도 될 것 같네요

Sooamazing replied:
아 postIntro가 게시글 발행 시 그 소개글인데, 사실 유저에게 받는 게 필수는 아니니까.. 근데? 아 지금 필수가 아니네요. ㅋㅋㅋㅋ 착각했어요. 너무 많은 걸 봐서 헷갈렸슴다.
우선, 저건 미리보기니까 프론트에서 글 내용 150자 뽑는 건 불가능할 것 같습니다. ㅋㅋㅋㅋㅋㅋ 
다음에 어떻게 했는지 보여주세요~ ㅋㅋㅋ

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/post/entity/Post.java:
```diff
@@ -0,0 +1,55 @@
+    private String postIntro;
+
+    @Column(nullable = false)
+    private Boolean isPublic = true;
```
오 생각해 보니 기본값을 여기서도 설정할 수 있네요. 전 항상 생성자나 정적팩토리 메서드로 만들어서 사용해서 거기서 설정했다 보니 완전 잊고 있었네요.. ㅋㅋ
(P3-답 궁금) 근데, DDL auto true면, 스키마에서의 기본값은 굳이 일부러 넣지 않은 건가요?

Yuurim98 replied:
넵 지금은 필드 초기값만으로도 충분할 것 같아서 넣지 않았습니다

Sooamazing replied:
오, 이것도 일종의 제약 조건 같다고 생각했는데요.(실제로 그런지 아닌지 모름)
아래의 상황도 고려해 보시면 좋을 것 같습니다.

nullable인데, default 값이 없어도 괜찮을까요? (스키마 상)
SQL 삽입 시, nullable만 있는 경우, 값을 지정하지 않고, default 값도 없으면 오류가 납니다. [참고1](https://velog.io/@codren/DB-not-null-%EA%B3%BC-default-%EB%91%98%EB%8B%A4-%EC%8D%A8%EC%95%BC%EB%90%98%EB%8A%94%EA%B0%80)

그리고, 참고로 만약 null 허용이었다면, default 값을 설정 안 했을 때 null로 들어가게 되는 거 알고 계셨나요?  MySQL에서는 boolean이 tinyint(1) 데이터 타입으로 지정되는 걸로 알고 있습니다. 이건 null이 기본값입니다.

위에 적어 놓았던 SQL문으로 삽입 등의 경우가 생각나서 말씀드려요. (이런 상황밖에 모르겠네요. ㅋㅋㅋ)

Yuurim98 replied:
값은 필수로 받아서 엔티티 필드에 채워줄 생각이긴 했었어서 말씀해주신 경우를 생각 못 했던 것 같아요
혹시 모를 상황을 대비해서 기본값을 지정해주는게 더 안전할 것 같네요 의견 감사합니다.

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/post/entity/Post.java:
```diff
@@ -0,0 +1,55 @@
+    @Column(nullable = false)
+    private String postUrl;
+
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private PostStatus postStatus = PostStatus.PUBLISHED;
```
(P3 - 답 궁금)
여기서 코드 단에서 Published로 기본값을 설정하신 이유가 궁금합니다. 
(P3 - 답 궁금) 임시 저장 글 테이블을 따로 분리한다고 하셨는데, 그냥 PostStatus로 구분하기로 변경하신 건가요?  원래는 제약 사항 등이 달라서 구분하는 게 맞는 것 같다고 하셨던 게 기억나서 왜인지 궁금해요.
(P5-그냥 궁금)
완전 억지 생각인 것 같은데, (적절한 예시도 생각 안 남) 혹시나 오류가 생겼을 경우, Published가 기본값이 되어도 괜찮을까요? IsPublic=true, Published 가 기본값이길래 궁금해졌습니다. 딱히 중요하지 않은 가정인 것 같기도 하지만요.

Yuurim98 replied:
임시저장 글을 따로 테이블로 분리했었는데 불필요할 것 같아서 그냥 게시글 테이블 컬럼으로 넣었고 필드 초기값을 설정한 건 공개 글은 상태값 없이 엔티티를 생성할 생각이었습니다.

그런데 의견주신 것처럼 갑자기 어떤 오류가 생길지 모르니 그냥 처음부터 명확한 값을 받아서 생성해주는게 좋을 것도 같아서 나중에는 수정될 수도 있을 것 같아요

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/post/entity/PostTag.java:
```diff
@@ -0,0 +1,28 @@
+import jakarta.persistence.ManyToOne;
+
+@Entity
+public class PostTag extends BaseEntity {
```
(P4-답 안 해도 됨. 근데 궁금.) 
Tag를 PostTag로 넣은 이유가 있는지 궁금합니다. Comment는 도메인으로 한 자리 차지 했는데, Tag는 Entity임에도 여기에 있길래요.
엥? 뭐야. Tag도 있고, PostTag도 있네요..? 왜죠? -> 아. 중간 테이블이군요.

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/post/entity/Post.java:
```diff
@@ -0,0 +1,55 @@
+    @Enumerated(EnumType.STRING)
+    private PostStatus postStatus = PostStatus.PUBLISHED;
+
+    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
+    private List<PostTag> postTags = new ArrayList<>();
+}
```
(P2-진지 검토 고려)
List로 구현하신 이유가 있을까요? Tag는 중복이 되면 안 되는데, 혹시 다 검사할 계획이셨는지 궁금합니다.
아니라면, 혹시 [데이터베이스/JPA
[JPA] 값 타입 컬렉션 : @ElementCollection, @CollectionTable](https://developer-hm.tistory.com/48) 이런 걸 들어 보셨나요? Set 자료구조에 사용할 수 있습니다. 저도 한 번 써 보고, 너무 오래돼서 여기에 적절한지 잘은 모르겠는데, 생명주기도 똑같이 가는 걸로 기억합니다. (아닐 수 있으니 자체 공부 필요함..)
Set을 고려했는데, List를 선택하신 거라면 이유가 궁금합니다. 아니라면, 한번 고려해 보실 수도 있을 것 같아 댓글 남깁니다.

Yuurim98 replied:
아 중복 불가 부분을 아예 고려를 안 하고 있었던 것 같습니다 수정하겠습니다.
그리고 공유주신 내용도 참고해서 반영할 수 있으면 해보겠습니다 감사합니다!

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/user/entity/User.java:
```diff
@@ -0,0 +1,40 @@
+    @Column(nullable = false)
+    private String password;
+
+    @Column(nullable = false)
```
(P5-그냥 궁금) 이렇게 하면, 다른 곳에서 참고할 때 열 이름이 user_id라 헷갈리지 않나요? 그냥 항상 제가 고민하던 거라 물어봅니다.

Yuurim98 replied:
헷갈릴 수도 있을 것 같긴한데 마땅한 이름이 생각이 안 나서 우선.. ㅋㅋㅋ

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/post/entity/Post.java:
```diff
@@ -0,0 +1,55 @@
+    @Column(nullable = false)
+    private String title;
+
+    private String content;
```
(P3-답 궁금)
여기는 Lob이 아닌 이유가 있나요? 제가 Lob에 대해 잘못 이해했을 수도...

Yuurim98 replied:
빼먹었네요,, 감사합니다

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/user/entity/User.java:
```diff
@@ -0,0 +1,40 @@
+    @Column(nullable = false)
+    private String blogName;
+
+    @Lob
+    private String profileIntro;
+
+    @Lob
+    private String blogIntro;
```
(P5-그냥 궁금) 여기가 그렇게 큰 용량이 필요할까요?

Yuurim98 replied:
이부분도 위에 내용과 같이 벨로그에 제한이 없어서 지정한 건데 여기도 고민해 볼 필요가 있겠네요

Sooamazing reviewed (commented):
오. 고생하셨습니다.
우선, 나름 꼼꼼히 열심히 보려고 했지만, 제가 사실 지금 기억이 아리까리해서.. ㅎㅎ.. 궁금한 내용 검색하면서 본 것들이라 트집 같은 이상한 질문이 있을 수 있다는 점 양해 부탁드립니다.. ㅎㅎ [뱅크 샐러드 P-n 룰](https://blog.banksalad.com/tech/banksalad-code-review-culture/)으로 질문의 강도(?)를 표기했으니 참고 부탁드립니다.
전체적으로 막 크게 이거 절대 안 됨. 이런 게 있는 게 아니라서 사실 질문은 그냥 질문으로만 남겨두셔도 될 것 같습니다. (전 영원히 궁금해 할 순 있어요... ^__^ㅋㅋㅋㅋㅋ)

## 패키지 구조
패키지 구조에 대해 궁금한 점이 있습니다. 
(P5-사소한 궁금증) 먼저 패키지 구조를 도메인 형식으로 선택하신 것 같은데, 선택하신 이유가 궁금합니다.
(P3-답 궁금) domain이 추가되지 않는 한 해당 depth의 directory는 추가될 일이 없없나요? ApiResponse, CustomException, config, 공통 코드 등이 생기면, 추후 common 아래에 두는 건지, 디렉토리가 따로 생기는지 등 구조 설계가 궁금합니다. (.)
아실 것 같긴 한데, 이런 식으로 섞어서도 사용하는 것 같아서 참고용 링크 첨부합니다. ([Spring Guide - Directory](https://cheese10yun.github.io/spring-guide-directory/#google_vignette))
좀 더 규모 있는 오픈 소스에서 찾아보려 했는데, 마땅치 않네요... ㅎㅎ.. 

## 버전 선택
(P5-사소한 궁금증) java, spring 버전은 어떤 걸로 택하셨는지, 왜 그걸 택했는지 궁금합니다. 

## SQL 스키마 제약
(P3-답 궁금)
전체적으로 Length, Unique 등은 따로 설정하지 않으신 것 같은데, 제약을 넣지 않은 이유가 궁금합니다.

## 구현
(P4-답 안 해도 됨. 근데 궁금.) postTags는 유일하게 Post 내부에 list로 존재하고, comment, series는 list로는 따로 구현하지 않은 이유가 있나요? 어떤 기준이 이런 차이를 만들었는지 궁금합니다.

azure1307 commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,53 @@
+       (1004, '맹구', 'maeng@example.com', 'user_4', '맹구 블로그', '안녕하세요', '환영합니다',
+        '^y4NQ!qxT#'),
+       (1005, '철수', 'chul@example.com', 'user_5', '철수 블로그', '^__^', null,
+        '$u03x@Vw0P');
```
(궁금) users 테이블의 user_id 컬럼은 언제 사용하나요? name, email, blog_name, profile_intro, password는 회원가입할 때 받는 컬럼같은데 user_id는 어떻게 사용하실건지 궁금합니다.

Yuurim98 replied:
벨로그 보니까 회원가입 할 때 받는 사용자 ID를 블로그 이름으로도 사용하고 식별자 ? 로도 사용하는 것 같더라고요 그래서 우선 해당 값을 식별자로 활용할 목적으로 추가해두었습니다.

azure1307 commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,53 @@
+VALUES (1001, 1001, 1002),
+       (1002, 1003, 1003),
+       (1003, 1005, 1003),
+       (1004, 1002, 1001);
```
(궁금) id 값이 다 1이 아니라 1001부터 시작하네요? 이렇게 하신 이유가 있는지 궁금합니다.!

Yuurim98 replied:
이유는 없고 그냥 예전에 했던 걸 참고하면서 한 건데 auto-increment 라서 1부터 시작했어도 상관 없었을 것 같습니다.. ㅎ

Yuurim98 commented:
@Sooamazing 

> 패키지 구조
> 패키지 구조에 대해 궁금한 점이 있습니다.
> (P5-사소한 궁금증) 먼저 패키지 구조를 도메인 형식으로 선택하신 것 같은데, 선택하신 이유가 궁금합니다.
* 각 도메인이 분리되어 있어서 직관적이고 유지보수에 용이하다고 생각했습니다.
> (P3-답 궁금) domain이 추가되지 않는 한 해당 depth의 directory는 추가될 일이 없없나요? ApiResponse, CustomException, config, 공통 코드 등이 생기면, 추후 common 아래에 두는 건지, 디렉토리가 따로 생기는지 등 구조 설계가 궁금합니다. (.)
* 예시로 들어주신 항목들 기준으로는 config를 제외한 나머지는 모두 common 하위에 둘 것 같습니다. config는 공통 설정이라기보다는 애플리케이션 전반에 영향을 부는 설정이라고 생각해서 별도 패키지로 분리하는게 더 자연스러울 것 같습니다.

> 버전 선택
> (P5-사소한 궁금증) java, spring 버전은 어떤 걸로 택하셨는지, 왜 그걸 택했는지 궁금합니다.
* java 21, spring 3.4.5 입니다. 특별한 이유는 없는데 나중에는 이런 부분도 생각해 보면 좋을 것 같네요

> SQL 스키마 제약
> (P3-답 궁금)
> 전체적으로 Length, Unique 등은 따로 설정하지 않으신 것 같은데, 제약을 넣지 않은 이유가 궁금합니다.
* ddl-auto는 개발 단계에서만 사용하기도 하고 어제 답변처럼 Validation 처리할 예정이었어서 제약조건을 굳이 달아야 하나 생각했던 것 같습니다.
> 구현
> (P4-답 안 해도 됨. 근데 궁금.) postTags는 유일하게 Post 내부에 list로 존재하고, comment, series는 list로는 따로 구현하지 않은 이유가 있나요? 어떤 기준이 이런 차이를 만들었는지 궁금합니다.
* PostTag는 게시글을 조회할 때 자주 함께 조회되는 연관 데이터라고 생각해서 리스트로 포함했고
Comment는 CommentRepository 등으로 별도로 조회할 생각이라서 리스트로 만들지 않았습니다.
또한 Series는 게시글이 하나의 시리즈에만 속한다고 생각해서 리스트로 관리할 필요가 없다고 판단했습니다.

yu-rim98 committed:
Refactor: deletedAt 삭제 #2

yu-rim98 committed:
Refactor: 타입, 제약 조건 수정 #2

엔티티 null 체크를 위해 NotNull 어노테이션 추가, 기본값 추가, List > Set 변경 및 equals, hashCode Override

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/tag/entity/Tag.java:
```diff
@@ -14,6 +16,24 @@ public class Tag extends BaseEntity {
+    @NotNull
+    @Column(length = 50, unique = true)
     private String name;
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Tag tag = (Tag) o;
+        return Objects.equals(id, tag.id) && Objects.equals(name, tag.name);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, name);
+    }
```
오. Unique 설정과 별개로 코드에서는 이렇게 검증해야 하는 건가요?

Yuurim98 replied:
DB 제약 조건과 별개로 코드 단에서 비교할 때 필요할 것 같아요

Sooamazing commented on src/main/java/com/diffbydevs/velog_clone/post/entity/Post.java:
```diff
@@ -33,23 +37,25 @@ public class Post extends BaseEntity {
 
     @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
-    private List<PostTag> postTags = new ArrayList<>();
+    private Set<PostTag> postTags = new HashSet<>();
```
오. 이런 글도 있긴 하네요... 머쓱. ㅎㅎ 
성능이 안 좋다면... 비교해 보고 알려주세요^_^ 하하.
[Collection type으로 Set 대신 List를 사용하는 이유가 있는지요?
](https://www.inflearn.com/community/questions/321256/collection-type%EC%9C%BC%EB%A1%9C-set-%EB%8C%80%EC%8B%A0-list%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0%EA%B0%80-%EC%9E%88%EB%8A%94%EC%A7%80%EC%9A%94?srsltid=AfmBOorbfZPR7lsyrkugwyYEoDrgn8ydGd2EeoOfO3JOpde5Tzoozt6M)

Yuurim98 replied:
오.. 알겠습니다 공유 감사합니다

Sooamazing reviewed (approved):
정상적으로 동작하는 거 확인됐으면 merge하셔도 좋을 것 같습니다~! 고생하셨습니다!

++
블로그에서 전체 태그를 확인하는 경우도 필요할 것 같습니다.

Yuurim98 commented:
@Sooamazing @azure1307 
전체 태그 조회할 때 User -> Post -> PostTag -> Tag 연관관계로 조인하면 조회할 수 있을 것 같아요.
![스크린샷 2025-05-09 오후 4 00 48](https://github.com/user-attachments/assets/92a953dd-4d1a-481b-b77e-47e8a3e6fe6d)
